File: packages/python/hushspec/extensions.py
"""Extension types for HushSpec: posture, origins, detection.

All types mirror the Rust and TypeScript implementations exactly, using
snake_case field names matching the YAML schema.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

from hushspec.rules import EgressRule, ToolAccessRule


# ============================================================
# Posture Extension
# ============================================================


class TransitionTrigger(str, Enum):
    """Event that triggers a posture state transition."""

    USER_APPROVAL = "user_approval"
    USER_DENIAL = "user_denial"
    CRITICAL_VIOLATION = "critical_violation"
    ANY_VIOLATION = "any_violation"
    TIMEOUT = "timeout"
    BUDGET_EXHAUSTED = "budget_exhausted"
    PATTERN_MATCH = "pattern_match"


@dataclass
class PostureState:
    """A named security state with capabilities and budgets."""

    description: Optional[str] = None
    capabilities: list[str] = field(default_factory=list)
    budgets: dict[str, int] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, data: dict) -> PostureState:
        return cls(
            description=data.get("description"),
            capabilities=list(data.get("capabilities", [])),
            budgets=dict(data.get("budgets", {})),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.description is not None:
            d["description"] = self.description
        d["capabilities"] = self.capabilities
        if self.budgets:
            d["budgets"] = self.budgets
        return d


@dataclass
class PostureTransition:
    """A transition rule between posture states.

    Uses ``from_state`` instead of ``from`` (Python keyword).
    Serialized as ``from`` in YAML.
    """

    from_state: str
    to: str
    on: TransitionTrigger
    after: Optional[str] = None

    @classmethod
    def from_dict(cls, data: dict) -> PostureTransition:
        # PyYAML interprets bare `on:` as boolean True (YAML 1.1 spec).
        # Look up both the string key "on" and the boolean key True.
        on_value = data.get("on", data.get(True))
        return cls(
            from_state=data["from"],
            to=data["to"],
            on=TransitionTrigger(on_value),
            after=data.get("after"),
        )

    def to_dict(self) -> dict:
        d: dict = {
            "from": self.from_state,
            "to": self.to,
            "on": self.on.value,
        }
        if self.after is not None:
            d["after"] = self.after
        return d


@dataclass
class PostureExtension:
    """Declarative state machine for capability and budget management."""

    initial: str
    states: dict[str, PostureState] = field(default_factory=dict)
    transitions: list[PostureTransition] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: dict) -> PostureExtension:
        states = {
            name: PostureState.from_dict(state_data)
            for name, state_data in data.get("states", {}).items()
        }
        transitions = [
            PostureTransition.from_dict(t) for t in data.get("transitions", [])
        ]
        return cls(
            initial=data["initial"],
            states=states,
            transitions=transitions,
        )

    def to_dict(self) -> dict:
        return {
            "initial": self.initial,
            "states": {name: state.to_dict() for name, state in self.states.items()},
            "transitions": [t.to_dict() for t in self.transitions],
        }


# ============================================================
# Origins Extension
# ============================================================


class OriginDefaultBehavior(str, Enum):
    """Behavior when no origin profile matches."""

    DENY = "deny"
    MINIMAL_PROFILE = "minimal_profile"


@dataclass
class OriginMatch:
    """Criteria for matching an origin context to a profile."""

    provider: Optional[str] = None
    tenant_id: Optional[str] = None
    space_id: Optional[str] = None
    space_type: Optional[str] = None
    visibility: Optional[str] = None
    external_participants: Optional[bool] = None
    tags: list[str] = field(default_factory=list)
    sensitivity: Optional[str] = None
    actor_role: Optional[str] = None

    @classmethod
    def from_dict(cls, data: dict) -> OriginMatch:
        return cls(
            provider=data.get("provider"),
            tenant_id=data.get("tenant_id"),
            space_id=data.get("space_id"),
            space_type=data.get("space_type"),
            visibility=data.get("visibility"),
            external_participants=data.get("external_participants"),
            tags=list(data.get("tags", [])),
            sensitivity=data.get("sensitivity"),
            actor_role=data.get("actor_role"),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.provider is not None:
            d["provider"] = self.provider
        if self.tenant_id is not None:
            d["tenant_id"] = self.tenant_id
        if self.space_id is not None:
            d["space_id"] = self.space_id
        if self.space_type is not None:
            d["space_type"] = self.space_type
        if self.visibility is not None:
            d["visibility"] = self.visibility
        if self.external_participants is not None:
            d["external_participants"] = self.external_participants
        if self.tags:
            d["tags"] = self.tags
        if self.sensitivity is not None:
            d["sensitivity"] = self.sensitivity
        if self.actor_role is not None:
            d["actor_role"] = self.actor_role
        return d


@dataclass
class OriginDataPolicy:
    """Data handling policy for an origin."""

    allow_external_sharing: bool = False
    redact_before_send: bool = False
    block_sensitive_outputs: bool = False

    @classmethod
    def from_dict(cls, data: dict) -> OriginDataPolicy:
        return cls(
            allow_external_sharing=data.get("allow_external_sharing", False),
            redact_before_send=data.get("redact_before_send", False),
            block_sensitive_outputs=data.get("block_sensitive_outputs", False),
        )

    def to_dict(self) -> dict:
        return {
            "allow_external_sharing": self.allow_external_sharing,
            "redact_before_send": self.redact_before_send,
            "block_sensitive_outputs": self.block_sensitive_outputs,
        }


@dataclass
class OriginBudgets:
    """Budget limits for an origin profile."""

    tool_calls: Optional[int] = None
    egress_calls: Optional[int] = None
    shell_commands: Optional[int] = None

    @classmethod
    def from_dict(cls, data: dict) -> OriginBudgets:
        return cls(
            tool_calls=data.get("tool_calls"),
            egress_calls=data.get("egress_calls"),
            shell_commands=data.get("shell_commands"),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.tool_calls is not None:
            d["tool_calls"] = self.tool_calls
        if self.egress_calls is not None:
            d["egress_calls"] = self.egress_calls
        if self.shell_commands is not None:
            d["shell_commands"] = self.shell_commands
        return d


@dataclass
class BridgeTarget:
    """A permitted cross-origin target."""

    provider: Optional[str] = None
    space_type: Optional[str] = None
    tags: list[str] = field(default_factory=list)
    visibility: Optional[str] = None

    @classmethod
    def from_dict(cls, data: dict) -> BridgeTarget:
        return cls(
            provider=data.get("provider"),
            space_type=data.get("space_type"),
            tags=list(data.get("tags", [])),
            visibility=data.get("visibility"),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.provider is not None:
            d["provider"] = self.provider
        if self.space_type is not None:
            d["space_type"] = self.space_type
        if self.tags:
            d["tags"] = self.tags
        if self.visibility is not None:
            d["visibility"] = self.visibility
        return d


@dataclass
class BridgePolicy:
    """Cross-origin transition control."""

    allow_cross_origin: bool = False
    allowed_targets: list[BridgeTarget] = field(default_factory=list)
    require_approval: bool = False

    @classmethod
    def from_dict(cls, data: dict) -> BridgePolicy:
        targets = [
            BridgeTarget.from_dict(t) for t in data.get("allowed_targets", [])
        ]
        return cls(
            allow_cross_origin=data.get("allow_cross_origin", False),
            allowed_targets=targets,
            require_approval=data.get("require_approval", False),
        )

    def to_dict(self) -> dict:
        d: dict = {
            "allow_cross_origin": self.allow_cross_origin,
            "require_approval": self.require_approval,
        }
        if self.allowed_targets:
            d["allowed_targets"] = [t.to_dict() for t in self.allowed_targets]
        return d


@dataclass
class OriginProfile:
    """An origin profile with match rules and security overrides.

    Uses ``match_rules`` instead of ``match`` (Python keyword).
    Serialized as ``match`` in YAML.
    """

    id: str
    match_rules: Optional[OriginMatch] = None
    posture: Optional[str] = None
    tool_access: Optional[ToolAccessRule] = None
    egress: Optional[EgressRule] = None
    data: Optional[OriginDataPolicy] = None
    budgets: Optional[OriginBudgets] = None
    bridge: Optional[BridgePolicy] = None
    explanation: Optional[str] = None

    @classmethod
    def from_dict(cls, data: dict) -> OriginProfile:
        return cls(
            id=data["id"],
            match_rules=(
                OriginMatch.from_dict(data["match"])
                if "match" in data
                else None
            ),
            posture=data.get("posture"),
            tool_access=(
                ToolAccessRule.from_dict(data["tool_access"])
                if "tool_access" in data
                else None
            ),
            egress=(
                EgressRule.from_dict(data["egress"])
                if "egress" in data
                else None
            ),
            data=(
                OriginDataPolicy.from_dict(data["data"])
                if "data" in data
                else None
            ),
            budgets=(
                OriginBudgets.from_dict(data["budgets"])
                if "budgets" in data
                else None
            ),
            bridge=(
                BridgePolicy.from_dict(data["bridge"])
                if "bridge" in data
                else None
            ),
            explanation=data.get("explanation"),
        )

    def to_dict(self) -> dict:
        d: dict = {"id": self.id}
        if self.match_rules is not None:
            d["match"] = self.match_rules.to_dict()
        if self.posture is not None:
            d["posture"] = self.posture
        if self.tool_access is not None:
            d["tool_access"] = self.tool_access.to_dict()
        if self.egress is not None:
            d["egress"] = self.egress.to_dict()
        if self.data is not None:
            d["data"] = self.data.to_dict()
        if self.budgets is not None:
            d["budgets"] = self.budgets.to_dict()
        if self.bridge is not None:
            d["bridge"] = self.bridge.to_dict()
        if self.explanation is not None:
            d["explanation"] = self.explanation
        return d


@dataclass
class OriginsExtension:
    """Origin-aware policy projection with match-based profiles."""

    default_behavior: Optional[OriginDefaultBehavior] = None
    profiles: list[OriginProfile] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: dict) -> OriginsExtension:
        default_behavior = None
        if "default_behavior" in data:
            default_behavior = OriginDefaultBehavior(data["default_behavior"])
        profiles = [OriginProfile.from_dict(p) for p in data.get("profiles", [])]
        return cls(
            default_behavior=default_behavior,
            profiles=profiles,
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.default_behavior is not None:
            d["default_behavior"] = self.default_behavior.value
        if self.profiles:
            d["profiles"] = [p.to_dict() for p in self.profiles]
        return d


# ============================================================
# Detection Extension
# ============================================================


class DetectionLevel(str, Enum):
    """Ordered severity level for detection results.

    The ordering is: SAFE < SUSPICIOUS < HIGH < CRITICAL.
    """

    SAFE = "safe"
    SUSPICIOUS = "suspicious"
    HIGH = "high"
    CRITICAL = "critical"

    def __lt__(self, other: object) -> bool:
        if not isinstance(other, DetectionLevel):
            return NotImplemented
        return _DETECTION_LEVEL_ORDER[self] < _DETECTION_LEVEL_ORDER[other]

    def __le__(self, other: object) -> bool:
        if not isinstance(other, DetectionLevel):
            return NotImplemented
        return _DETECTION_LEVEL_ORDER[self] <= _DETECTION_LEVEL_ORDER[other]

    def __gt__(self, other: object) -> bool:
        if not isinstance(other, DetectionLevel):
            return NotImplemented
        return _DETECTION_LEVEL_ORDER[self] > _DETECTION_LEVEL_ORDER[other]

    def __ge__(self, other: object) -> bool:
        if not isinstance(other, DetectionLevel):
            return NotImplemented
        return _DETECTION_LEVEL_ORDER[self] >= _DETECTION_LEVEL_ORDER[other]


_DETECTION_LEVEL_ORDER = {
    DetectionLevel.SAFE: 0,
    DetectionLevel.SUSPICIOUS: 1,
    DetectionLevel.HIGH: 2,
    DetectionLevel.CRITICAL: 3,
}


@dataclass
class PromptInjectionDetection:
    """Prompt injection detection thresholds."""

    enabled: Optional[bool] = None
    warn_at_or_above: Optional[DetectionLevel] = None
    block_at_or_above: Optional[DetectionLevel] = None
    max_scan_bytes: Optional[int] = None

    @classmethod
    def from_dict(cls, data: dict) -> PromptInjectionDetection:
        return cls(
            enabled=data.get("enabled"),
            warn_at_or_above=(
                DetectionLevel(data["warn_at_or_above"])
                if "warn_at_or_above" in data
                else None
            ),
            block_at_or_above=(
                DetectionLevel(data["block_at_or_above"])
                if "block_at_or_above" in data
                else None
            ),
            max_scan_bytes=data.get("max_scan_bytes"),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.enabled is not None:
            d["enabled"] = self.enabled
        if self.warn_at_or_above is not None:
            d["warn_at_or_above"] = self.warn_at_or_above.value
        if self.block_at_or_above is not None:
            d["block_at_or_above"] = self.block_at_or_above.value
        if self.max_scan_bytes is not None:
            d["max_scan_bytes"] = self.max_scan_bytes
        return d


@dataclass
class JailbreakDetection:
    """Jailbreak detection thresholds."""

    enabled: Optional[bool] = None
    block_threshold: Optional[int] = None
    warn_threshold: Optional[int] = None
    max_input_bytes: Optional[int] = None

    @classmethod
    def from_dict(cls, data: dict) -> JailbreakDetection:
        return cls(
            enabled=data.get("enabled"),
            block_threshold=data.get("block_threshold"),
            warn_threshold=data.get("warn_threshold"),
            max_input_bytes=data.get("max_input_bytes"),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.enabled is not None:
            d["enabled"] = self.enabled
        if self.block_threshold is not None:
            d["block_threshold"] = self.block_threshold
        if self.warn_threshold is not None:
            d["warn_threshold"] = self.warn_threshold
        if self.max_input_bytes is not None:
            d["max_input_bytes"] = self.max_input_bytes
        return d


@dataclass
class ThreatIntelDetection:
    """Threat intelligence screening configuration."""

    enabled: Optional[bool] = None
    pattern_db: Optional[str] = None
    similarity_threshold: Optional[float] = None
    top_k: Optional[int] = None

    @classmethod
    def from_dict(cls, data: dict) -> ThreatIntelDetection:
        return cls(
            enabled=data.get("enabled"),
            pattern_db=data.get("pattern_db"),
            similarity_threshold=data.get("similarity_threshold"),
            top_k=data.get("top_k"),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.enabled is not None:
            d["enabled"] = self.enabled
        if self.pattern_db is not None:
            d["pattern_db"] = self.pattern_db
        if self.similarity_threshold is not None:
            d["similarity_threshold"] = self.similarity_threshold
        if self.top_k is not None:
            d["top_k"] = self.top_k
        return d


@dataclass
class DetectionExtension:
    """Detection engine threshold configuration."""

    prompt_injection: Optional[PromptInjectionDetection] = None
    jailbreak: Optional[JailbreakDetection] = None
    threat_intel: Optional[ThreatIntelDetection] = None

    @classmethod
    def from_dict(cls, data: dict) -> DetectionExtension:
        return cls(
            prompt_injection=(
                PromptInjectionDetection.from_dict(data["prompt_injection"])
                if "prompt_injection" in data
                else None
            ),
            jailbreak=(
                JailbreakDetection.from_dict(data["jailbreak"])
                if "jailbreak" in data
                else None
            ),
            threat_intel=(
                ThreatIntelDetection.from_dict(data["threat_intel"])
                if "threat_intel" in data
                else None
            ),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.prompt_injection is not None:
            d["prompt_injection"] = self.prompt_injection.to_dict()
        if self.jailbreak is not None:
            d["jailbreak"] = self.jailbreak.to_dict()
        if self.threat_intel is not None:
            d["threat_intel"] = self.threat_intel.to_dict()
        return d


# ============================================================
# Extensions Container
# ============================================================


@dataclass
class Extensions:
    """Optional extension modules for advanced features."""

    posture: Optional[PostureExtension] = None
    origins: Optional[OriginsExtension] = None
    detection: Optional[DetectionExtension] = None

    @classmethod
    def from_dict(cls, data: dict) -> Extensions:
        return cls(
            posture=(
                PostureExtension.from_dict(data["posture"])
                if "posture" in data
                else None
            ),
            origins=(
                OriginsExtension.from_dict(data["origins"])
                if "origins" in data
                else None
            ),
            detection=(
                DetectionExtension.from_dict(data["detection"])
                if "detection" in data
                else None
            ),
        )

    def to_dict(self) -> dict:
        d: dict = {}
        if self.posture is not None:
            d["posture"] = self.posture.to_dict()
        if self.origins is not None:
            d["origins"] = self.origins.to_dict()
        if self.detection is not None:
            d["detection"] = self.detection.to_dict()
        return d

File: crates/hushspec/src/extensions.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

use crate::rules::{EgressRule, ToolAccessRule};

// --- Top-level Extensions container ---

/// Optional extension modules for advanced features.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Extensions {
    /// Declarative posture state machine.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub posture: Option<PostureExtension>,
    /// Origin-aware policy projection.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub origins: Option<OriginsExtension>,
    /// Detection engine threshold configuration.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detection: Option<DetectionExtension>,
}

// --- Posture Extension ---

/// Declarative state machine for capability and budget management.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PostureExtension {
    /// Name of the starting state.
    pub initial: String,
    /// Named security states with capabilities and budgets.
    pub states: BTreeMap<String, PostureState>,
    /// Rules governing transitions between states.
    #[serde(default)]
    pub transitions: Vec<PostureTransition>,
}

/// A named security state with capabilities and budgets.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PostureState {
    /// Human-readable description of this state.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Capabilities available in this state.
    #[serde(default)]
    pub capabilities: Vec<String>,
    /// Resource budgets (e.g. `"tool_calls": 100`). Must be non-negative.
    #[serde(default)]
    pub budgets: BTreeMap<String, i64>,
}

/// A transition rule between posture states.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PostureTransition {
    /// Source state name, or `"*"` for any state.
    pub from: String,
    /// Target state name.
    pub to: String,
    /// Event that triggers this transition.
    pub on: TransitionTrigger,
    /// Duration string required for `Timeout` triggers (e.g. `"30s"`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub after: Option<String>,
}

/// Event that triggers a posture state transition.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransitionTrigger {
    /// Operator approved the current action.
    UserApproval,
    /// Operator denied the current action.
    UserDenial,
    /// A critical-severity violation occurred.
    CriticalViolation,
    /// Any violation occurred.
    AnyViolation,
    /// A duration elapsed (requires `after` field).
    Timeout,
    /// A budget counter reached zero.
    BudgetExhausted,
    /// A threat-intel pattern matched.
    PatternMatch,
}

// --- Origins Extension ---

/// Origin-aware policy projection with match-based profiles.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OriginsExtension {
    /// Behavior when no origin profile matches.
    #[serde(default)]
    pub default_behavior: OriginDefaultBehavior,
    /// Ordered list of origin profiles; first match wins.
    #[serde(default)]
    pub profiles: Vec<OriginProfile>,
}

/// Behavior when no origin profile matches.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OriginDefaultBehavior {
    /// Deny all actions from unmatched origins.
    #[default]
    Deny,
    /// Apply a minimal-privilege profile.
    MinimalProfile,
}

/// An origin profile with match rules and security overrides.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OriginProfile {
    /// Unique identifier for this profile.
    pub id: String,
    /// Criteria for matching an origin context to this profile.
    #[serde(default, rename = "match", skip_serializing_if = "Option::is_none")]
    pub match_rules: Option<OriginMatch>,
    /// Posture state to activate for this origin.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub posture: Option<String>,
    /// Tool access overrides for this origin.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_access: Option<ToolAccessRule>,
    /// Egress overrides for this origin.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub egress: Option<EgressRule>,
    /// Data handling policy for this origin.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub data: Option<OriginDataPolicy>,
    /// Budget limits for this origin.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub budgets: Option<OriginBudgets>,
    /// Cross-origin transition controls.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bridge: Option<BridgePolicy>,
    /// Human-readable rationale for this profile.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub explanation: Option<String>,
}

/// Criteria for matching an origin context to a profile.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OriginMatch {
    /// Platform provider (e.g. `"slack"`, `"github"`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// Tenant/organization identifier.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tenant_id: Option<String>,
    /// Specific space/channel identifier.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub space_id: Option<String>,
    /// Space type (e.g. `"channel"`, `"dm"`, `"thread"`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub space_type: Option<String>,
    /// Visibility level (e.g. `"public"`, `"private"`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub visibility: Option<String>,
    /// Whether external participants are present.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub external_participants: Option<bool>,
    /// Required tags on the origin context.
    #[serde(default)]
    pub tags: Vec<String>,
    /// Data sensitivity classification.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sensitivity: Option<String>,
    /// Required actor role.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub actor_role: Option<String>,
}

/// Data handling policy for an origin.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OriginDataPolicy {
    /// Whether data may be shared with external parties.
    #[serde(default)]
    pub allow_external_sharing: bool,
    /// Whether to redact sensitive content before sending.
    #[serde(default)]
    pub redact_before_send: bool,
    /// Whether to suppress sensitive outputs entirely.
    #[serde(default)]
    pub block_sensitive_outputs: bool,
}

/// Budget limits for an origin profile.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OriginBudgets {
    /// Maximum tool invocations allowed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<u64>,
    /// Maximum network egress calls allowed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub egress_calls: Option<u64>,
    /// Maximum shell commands allowed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub shell_commands: Option<u64>,
}

/// Cross-origin transition control.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BridgePolicy {
    /// Whether cross-origin transitions are permitted.
    #[serde(default)]
    pub allow_cross_origin: bool,
    /// Permitted cross-origin targets.
    #[serde(default)]
    pub allowed_targets: Vec<BridgeTarget>,
    /// Whether cross-origin transitions require operator approval.
    #[serde(default)]
    pub require_approval: bool,
}

/// A permitted cross-origin target.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BridgeTarget {
    /// Target platform provider.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    /// Target space type.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub space_type: Option<String>,
    /// Required tags on the target.
    #[serde(default)]
    pub tags: Vec<String>,
    /// Required visibility level on the target.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub visibility: Option<String>,
}

// --- Detection Extension ---

/// Detection engine threshold configuration.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DetectionExtension {
    /// Prompt injection detection thresholds.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prompt_injection: Option<PromptInjectionDetection>,
    /// Jailbreak detection thresholds.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub jailbreak: Option<JailbreakDetection>,
    /// Threat intelligence screening configuration.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub threat_intel: Option<ThreatIntelDetection>,
}

/// Prompt injection detection thresholds.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PromptInjectionDetection {
    /// Whether prompt injection detection is active. Defaults to `true`.
    #[serde(default = "crate::rules::default_true")]
    pub enabled: bool,
    /// Detection level at or above which a warning is emitted.
    #[serde(default = "default_suspicious")]
    pub warn_at_or_above: DetectionLevel,
    /// Detection level at or above which the action is blocked.
    #[serde(default = "default_high")]
    pub block_at_or_above: DetectionLevel,
    /// Maximum input bytes to scan. Defaults to 200,000.
    #[serde(default = "default_scan_bytes")]
    pub max_scan_bytes: usize,
}

/// Ordered severity level for detection results.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DetectionLevel {
    /// No threat detected.
    Safe,
    /// Possible threat; warrants review.
    Suspicious,
    /// Likely threat; should block by default.
    High,
    /// Definite threat; must block.
    Critical,
}

/// Jailbreak detection thresholds.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct JailbreakDetection {
    /// Whether jailbreak detection is active. Defaults to `true`.
    #[serde(default = "crate::rules::default_true")]
    pub enabled: bool,
    /// Score at or above which the action is blocked. Defaults to 70.
    #[serde(default = "default_block_threshold")]
    pub block_threshold: u32,
    /// Score at or above which a warning is emitted. Defaults to 30.
    #[serde(default = "default_warn_threshold")]
    pub warn_threshold: u32,
    /// Maximum input bytes to scan. Defaults to 200,000.
    #[serde(default = "default_scan_bytes")]
    pub max_input_bytes: usize,
}

/// Threat intelligence screening configuration.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ThreatIntelDetection {
    /// Whether threat intel screening is active. Defaults to `true`.
    #[serde(default = "crate::rules::default_true")]
    pub enabled: bool,
    /// Path to the pattern database, or `builtin:` prefix for embedded patterns.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pattern_db: Option<String>,
    /// Cosine similarity threshold for matching. Defaults to 0.85.
    #[serde(default = "default_similarity_threshold")]
    pub similarity_threshold: f64,
    /// Number of top matches to return. Defaults to 5.
    #[serde(default = "default_top_k")]
    pub top_k: usize,
}

// Default helpers
pub(crate) fn default_suspicious() -> DetectionLevel {
    DetectionLevel::Suspicious
}
pub(crate) fn default_high() -> DetectionLevel {
    DetectionLevel::High
}
pub(crate) fn default_scan_bytes() -> usize {
    200_000
}
pub(crate) fn default_block_threshold() -> u32 {
    70
}
pub(crate) fn default_warn_threshold() -> u32 {
    30
}
pub(crate) fn default_similarity_threshold() -> f64 {
    0.85
}
pub(crate) fn default_top_k() -> usize {
    5
}
